File: src/code/python/sorting_algorithms/bucket_sort.py
def bucket_sort(arr: list) -> list:
    num_buckets = 10
    min_num = min(arr)
    max_num = max(arr)
    bucket_size = (max_num - min_num) / num_buckets
    buckets = [[] for _ in range(num_buckets)]

    for num in arr:
        index = min(int((num - min_num) / bucket_size), num_buckets - 1)
        buckets[index].append(num)

    return [num for bucket in buckets for num in sorted(bucket)]

File: src/code/ruby/sorting_algorithms/bucket_sort.rb
def bucket_sort(arr)
  num_buckets = 10
  min_num     = arr.min
  max_num     = arr.max
  bucket_size = (max_num - min_num).to_f / num_buckets
  buckets     = Array.new(num_buckets) { [] }

  arr.each do |num|
    index = [( (num - min_num) / bucket_size ).to_i, num_buckets - 1].min
    buckets[index] << num
  end

  buckets.map { |bucket| bucket.sort }.flatten
end
